### PR TITLE
Chairman

### PR DESCRIPTION
**Pull request by coot**
Example usage, run the demo and in another terminal:
```
cabal new-run chairman -- -n 0 -n 1 -n 2 --real-pbft -k 5 -s 20 -t 40
```
This will evaluate first 20 blocks, and it will timeout after 40 seconds. (Currently this will timeout as the demo is not working)

or run the demo with `--bft` and
```
 cabal new-run chairman -- -n 0 -n 1 -n 2 --bft -k 5
```
it will continuously track if there is consensus (`--bft` runs fine).

### CODE REVIEW

File: chairman/Chairman.hs [hunk 1]
```diff
@@ -0,0 +1,434 @@
+{-# LANGUAGE BangPatterns        #-}
+{-# LANGUAGE DataKinds           #-}
+{-# LANGUAGE FlexibleContexts    #-}
+{-# LANGUAGE GADTs               #-}
+{-# LANGUAGE KindSignatures      #-}
+{-# LANGUAGE NamedFieldPuns      #-}
```
Comment (erikd) [inline chairman/Chairman.hs]: :tada:

File: chairman/Chairman.hs [hunk 2]
```diff
@@ -0,0 +1,434 @@
...
+                      ( NotFoundCommonBlock tips
+                      , chains'
+                      )
+                    -- there is an intersection and all forks are short
+                    Just headSlot ->
+                      ( WitnessedConsensusAt headSlot (map Block.blockPoint $ catMaybes tips)
+                      , chains'
+                      )
+            -- there is a long fork
+            else throw $ NodeMisconduct tips
```
Comment (erikd) [inline chairman/Chairman.hs]: What monad is this throwing in?

File: chairman/Main.hs [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+      (Just timeout, SomeProtocol protocol) ->
+        runChairman protocol
+                    caCoreNodeIds
+                    (NumCoreNodes $ length caCoreNodeIds)
+                    caSecurityParam
+                    caMaxBlockNo
+                    stdoutTracer
+        `race_` do
+          threadDelay (timeout * 1_000_000)
+          throwIO Timeout
```
Comment (erikd) [inline chairman/Main.hs]: At least here I know its `IO`.

File: chairman/Chairman.hs [hunk 4]
```diff
@@ -0,0 +1,434 @@
...
+instance Condense blk => Show (ChairmanError blk) where
+    show (NodeMisconduct blks) = "NodeMisconduct " ++ condense blks
+
+instance (Condense blk, Typeable blk) => Exception (ChairmanError blk)
+
+
+-- | Check if the oldest immutable tip agree, keep forks from the last common
+-- block (including it).
+--
+checkAndSubtract
```
Comment (mrBliss) [inline chairman/Chairman.hs]: `Subtract` here is a bit confusing (in the helper too). Isn't there a clearer name?

File: chairman/Chairman.hs (hunk 2 above)
Comment (coot) [reply chairman/Chairman.hs]: here it throws in pure code, it's wrapped in `MonadSTM m`, which at runtime will be `STM`.

Comment (mrBliss) [reply chairman/Chairman.hs]: There is a comment `it isn't really pure as it throws the 'NodeMisconduct' exception`. I would advise using `Except` or `Either` for this function and then throwing the exception in `STM` (use `MonadThrow (STM m`).

File: chairman/Chairman.hs [hunk 5]
```diff
@@ -0,0 +1,434 @@
...
+    return res
+  where
+    -- it isn't really pure as it throws the 'NodeMisconduct' exception
+    subtractPure :: Map CoreNodeId (ChainFragment blk)
+                 -> ( ChairmanTrace blk
+                    , Map CoreNodeId (ChainFragment blk)
+                    )
+    subtractPure chains =
+      let tips :: [Maybe blk]
+          tips = foldl' (\tips' cf -> CF.head cf : tips') [] chains
```
Comment (mrBliss) [inline chairman/Chairman.hs]: This is just `map` in disguise, I would prefer:
```haskell
tips = map CF.head $ Map.elems chains
```

File: chairman/Chairman.hs [hunk 6]
```diff
@@ -0,0 +1,434 @@
...
+          -- find intersection of all the chains; the CLI guarantees that
+          -- there's at least one entry in @chains@, thus @fold1@ is safe here.
+          common :: ChainFragment blk
+          common =
+            foldl1
+              (\common' cf ->
+                case CF.intersectChainFragments common' cf of
+                  Nothing -> CF.Empty
+                  Just (_, common'', _, _)
+                          -> common'')
```
Comment (mrBliss) [inline chairman/Chairman.hs]: Having this on a separate line just looks weird to me :slightly_smiling_face:. I would put it on the previous line and align the `->`

File: chairman/Chairman.hs [hunk 7]
```diff
@@ -0,0 +1,434 @@
...
+              chains
+
+          mbHeadSlot = CF.headSlot common
```
Comment (mrBliss) [inline chairman/Chairman.hs]: `mbCommonHeadSlot`?

See the next comment: use a `Point` instead!

File: chairman/Chairman.hs [hunk 8]
```diff
@@ -0,0 +1,434 @@
...
+      txSubmissionClient txv
+
+--
+-- Chain-Sync client
+--
+
+-- | 'ChainSyncClient' which build chain fragment; on every roll forward it will
+-- check if there is consensus on immutable chain.
+--
+chainSyncClient
```
Comment (mrBliss) [inline chairman/Chairman.hs]: Can you move this up a bit, closer to the implementation of the helper functions it uses (`addBlock`, etc.)? Because I constantly had to scroll back and forth over unrelated code :slightly_smiling_face:

File: chairman/Chairman.hs (hunk 2 above)
Comment (coot) [reply chairman/Chairman.hs]: Done.

File: cardano-node/Cardano/Node/CLI.hs [hunk 9]
```diff
@@ -113,7 +113,7 @@ fromProtocol RealPBFT =
     case Consensus.runProtocol p of
       Dict -> return $ SomeProtocol p
   where
-    p = ProtocolRealPBFT defaultDemoPBftParams genesisConfig
+    p = ProtocolRealPBFT defaultDemoPBftParams genesisConfig Nothing
```
[the remark this answers is not in the document]
Comment (coot) [reply cardano-node/Cardano/Node/CLI.hs]: Right, we need to merge #60 first.

File: chairman/Main.hs [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+    ChairmanArgs { caProtocol
+                 , caCoreNodeIds
+                 , caSecurityParam
+                 , caMaxBlockNo
+                 , caTimeout
+                 } <- execParser opts
+    p <- fromProtocol caProtocol
+    case (caTimeout, p) of
+      (Nothing, SomeProtocol protocol) ->
+        runChairman protocol
```
[the remark this answers is not in the document]
Comment (coot) [reply chairman/Main.hs]: Done

File: chairman/Main.hs [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+
+import           Cardano.Node.CLI
+
+import           Chairman (runChairman)
+
+
+data ChairmanArgs = ChairmanArgs {
+      caProtocol      :: Protocol
+    , caCoreNodeIds   :: [CoreNodeId]
+    , caSecurityParam :: Int
```
[the remark this answers is not in the document]
Comment (coot) [reply chairman/Main.hs]: Done

File: chairman/Chairman.hs [hunk 12]
```diff
@@ -0,0 +1,434 @@
...
+                           ptcl
+
+        in connectTo
+            (,)
+            (localInitiatorNetworkApplication
+              coreNodeId
+              chainsVar
+              securityParam
+              maxBlockNo
+              (contramap show tracer)
```
[the remark this answers is not in the document]
Comment (coot) [reply chairman/Chairman.hs]: Nice, I wasn't aware.

File: chairman/Chairman.hs (hunk 7 above)
Comment (coot) [reply chairman/Chairman.hs]: > See the next comment: use a Point instead!

I don't see it; Either will work, because the slot is occupied by the same block in each chain.  But using `Point` might be more instructive.

File: chairman/Chairman.hs (hunk 6 above)
Comment (coot) [reply chairman/Chairman.hs]: Right, it was much longer in a previous version...

**Comment by coot** [comment]
@mrBliss I pushed fixes for all your comments, except breaking the `demo`. Let's merge #60 first.

**Comment by mrBliss** [comment]
> @mrBliss I pushed fixes for all your comments, except breaking the `demo`. Let's merge #60 first.

There a still some unresolved comments, I think because GitHub collapsed them. There should be a button "7 hidden conversations **Load more...**"

**Comment by coot** [comment]
Are you sure, the only non comment which I haven't addressed is the one which breaks the `demo`, but I prefer to wait for #60 before merging this one.

**Comment by mrBliss** [comment]
@coot:
https://github.com/input-output-hk/cardano-node/pull/80#discussion_r307648434
https://github.com/input-output-hk/cardano-node/pull/80#discussion_r307648963
https://github.com/input-output-hk/cardano-node/pull/80#discussion_r307650188
https://github.com/input-output-hk/cardano-node/pull/80#discussion_r307650463
https://github.com/input-output-hk/cardano-node/pull/80#discussion_r307651334
https://github.com/input-output-hk/cardano-node/pull/80#discussion_r307651714

File: chairman/Chairman.hs [hunk 13]
```diff
@@ -0,0 +1,434 @@
...
+--
+-- Submission client
+--
+
+-- | A 'LocalTxSubmissionClient' that submits transactions reading them from
+-- a 'TMVar'.  A real implementation should use a better synchronisation
+-- primitive.  This demo creates and empty 'TMVar' in
+-- 'muxLocalInitiatorNetworkApplication' above and never fills it with a tx.
+--
+txSubmissionClient
```
[the remark this answers is not in the document]
Comment (coot) [reply chairman/Chairman.hs]: Because local client needs both chain sync and tx submission. It's a limitation of the `ouroboros-network` interface, and partially a limitation of the multiplexer (if any mini-protocol exists, the multiplexer will kill all other mini-protocol threads as well).

File: chairman/Chairman.hs [hunk 14]
```diff
@@ -0,0 +1,434 @@
...
+import           Ouroboros.Consensus.Protocol
+import           Ouroboros.Consensus.Util.Condense
+
+import           Network.TypedProtocol.Codec
+import           Network.TypedProtocol.Codec.Cbor
+import           Network.TypedProtocol.Driver
+import           Ouroboros.Network.Mux
+import           Ouroboros.Network.Block (BlockNo, HasHeader, HeaderHash, Point, SlotNo)
+import qualified Ouroboros.Network.Block as Block
+import           Ouroboros.Network.ChainFragment (ChainFragment)
```
[the remark this answers is not in the document]
Comment (coot) [reply chairman/Chairman.hs]: but `AnchoredFragment` requires that I know the `anchorPoint`; I start with an empty chain, which is not anchored, then I will have to use `Maybe AnchoredFragment`. I feel that I will only have additional layer of `Maybe`s to deal with.

File: chairman/Chairman.hs [hunk 15]
```diff
@@ -0,0 +1,434 @@
...
+       )
+    => CoreNodeId
+    -> ChainsVar m blk
+    -> blk
+    -> STM m ()
+addBlock coreNodeId chainsVar blk =
+    modifyTVar' chainsVar (Map.adjust (CF.addBlock blk) coreNodeId)
+
+
+data ChairmanError blk = NodeMisconduct [blk]
```
[the remark this answers is not in the document]
Comment (coot) [reply chairman/Chairman.hs]: I'll add a comment

File: chairman/Chairman.hs (hunk 14 above)
[the remark this answers is not in the document]
Comment (mrBliss) [reply chairman/Chairman.hs]: The `anchorPoint` of an empty chain is ... genesis (`GenesisPoint`) 🙂. No `Maybe`s needed.

File: chairman/Chairman.hs (hunk 13 above)
[the remark this answers is not in the document]
Comment (mrBliss) [reply chairman/Chairman.hs]: I see, that's unfortunate.

File: chairman/Chairman.hs [hunk 16]
```diff
@@ -0,0 +1,434 @@
...
+              chains
+
+          mbHeadSlot = CF.headSlot common
+
+          -- remove common chain fragment from all the chains, but preserve the
+          -- common intersection point, this way we avoid leaving empty chains.
+          !chains'  = case mbHeadSlot of
+            Nothing       -> chains
+            Just headSlot ->
+                  (CF.takeWhileNewest (\blk -> Block.blockSlot blk >= headSlot))
```
[the remark this answers is not in the document]
Comment (coot) [reply chairman/Chairman.hs]: Good point, I'll use a point here.

File: chairman/Chairman.hs (hunk 4 above)
Comment (coot) [reply chairman/Chairman.hs]: what about `prune`?

Comment (mrBliss) [reply chairman/Chairman.hs]: Yeah, that's better 👍

**Comment by coot** [comment]
@mrBliss done, using `AnchoredChains` actually simplified things. I also added a commit which handles `MuxError` thrown when reading from closed bearer.

File: cardano-node/Cardano/Node/CLI.hs (hunk 9 above)
[the remark this answers is not in the document]
Comment (coot) [reply cardano-node/Cardano/Node/CLI.hs]: Fixed

File: chairman/Chairman.hs [hunk 17]
```diff
@@ -0,0 +1,468 @@
...
+      traceWith tracer (show err)
+      atomically $ modifyTVar' chainsVar (Map.delete coreNodeId)
+
+
+data ChairmanTrace blk
+  = NotFoundCommonBlock [Point blk]
+  -- ^ the common block was not found, present list of head points.  Might be
+  -- empty if the chain was empty.
+  | WitnessedConsensusAt (Point blk) [Point blk]
+  -- ^ witness consenss at a given point.  The list is a list of tip points of
```
Comment (mrBliss) [inline chairman/Chairman.hs]: `s/consenss/consensus/`

File: chairman/Chairman.hs [hunk 18]
```diff
@@ -0,0 +1,468 @@
...
+    -> STM m (ChairmanTrace blk)
+checkAndPrune chainsVar (SecurityParam securityParam) = do
+    chains <- readTVar chainsVar
+    case checkAndPrunePure chains of
+      Left err -> throwM err
+      Right (res, chains') -> do
+        writeTVar chainsVar $! chains'
+        return res
+  where
+    -- it isn't really pure as it throws the 'NodeMisconduct' exception
```
Comment (mrBliss) [inline chairman/Chairman.hs]: This comment is outdated.

File: chairman/Chairman.hs (hunk 17 above)
Comment (coot) [reply chairman/Chairman.hs]: Good catch

**Comment by coot** [comment]
Done, and rebased on top of merged #60.  I couldn't check weather it works since `script/demo.sh` got broken.
